src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Options/PackageSourcesPage.cs

```diff
@@ -0,0 +1,243 @@
+        {
+            ExternalSettingOperationResult result;
+
+            try
+            {
+                List<PackageSource> packageSources = new List<PackageSource>(capacity: packageSourceDictionaryList.Count);
+                List<PackageSource> existingPackageSources = LoadPackageSources(isMachineWide: false);
+
+                foreach (Dictionary<string, object> packageSourceDictionary in packageSourceDictionaryList)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    string name = packageSourceDictionary[MonikerSourceName].ToString();
+                    string source = packageSourceDictionary[MonikerSourceUrl].ToString();
+                    bool isEnabled = (bool)packageSourceDictionary[MonikerIsEnabled];
+
+                    PackageSource packageSource =
+                        PackageSourceValidator.FindExistingOrCreate(
+                            source,
+                            name,
+                            isEnabled,
+                            existingPackageSources);
+
+                    packageSources.Add(packageSource);
+                }
+
+                // Throw any validation errors before saving.
+                PackageSourceValidator.ValidateUniquenessOrThrow(packageSources);
+
+                _packageSourceProvider.SavePackageSources(packageSources);
+                result = ExternalSettingOperationResult.Success.Instance;
+            }
+            catch (OperationCanceledException ex) when (cancellationToken.IsCancellationRequested)
+            {
+                throw ex;
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
+            catch (Exception ex)
+#pragma warning restore CA1031 // Do not catch general exception types
+            {
+                result = CreateSettingErrorResult(ex.Message);
+                ActivityLog.LogError(ExceptionHelper.LogEntrySource, ex.ToString());
+            }
+
+            return result;
+        }
+
+        private static ExternalSettingOperationResult<T> GetValuePackageSources<T>(List<PackageSource> packageSources)
+        {
+            ExternalSettingOperationResult<T> result;
+
+            try
+            {
+                var packageSourcesList = new List<Dictionary<string, object>>(capacity: packageSources.Count);
+
+                // Each list item is represented by a dictionary, which in this case will have a single key-value pair for ConfigPath.
+                foreach (PackageSource packageSource in packageSources)
+                {
+                    var dict = new Dictionary<string, object>(capacity: 3)
+                    {
+                        { MonikerSourceName, packageSource.Name },
+                        { MonikerSourceUrl, packageSource.SourceUri }, // Throws if Source is an invalid URI
```

Comment:
I considered using `TrySourceUri`, but I'd rather the initial page loading fail here when there's an invalid nuget.config.
Once the customer finds the page loaded, I don't want it to be in an invalid state which only errors after they make a change.
Future iterations can improve this as we get more validation support from USX.